### PR TITLE
Extend refresh exception handling to non-standard and system exceptions

### PR DESCRIPTION
**Pull request by saintx**
Extends exception handling on refresh to non-standard, system exceptions.

### CODE REVIEW

File: memoize/wrapper.py [hunk 1]
```diff
@@ -96,6 +96,9 @@ async def refresh(actual_entry: Optional[CacheEntry], key: CacheKey,
         except Exception as e:
             logger.debug('Early failure instantiating coroutine for %s: %s', key, e)
             raise CachedMethodFailedException('Refresh failed to start') from e
+        except BaseException as be:
+            logger.debug("Non-standard exception caught during coroutine instantiation for %s: %s", key, be)
+            raise CachedMethodFailedException('Refresh failed to start') from be
```
Comment (charles-dyfis-net) [inline memoize/wrapper.py]: I'd probably _only_ catch BaseException here -- an invocation-time vs execution-time failure is a corner case to begin with.

Comment (saintx) [reply memoize/wrapper.py]: Done, updated the areas where we _catch or check for `isinstance` of_ `Exception` to `BaseException`. That way we're being comprehensive.

File: memoize/wrapper.py [hunk 2]
```diff
@@ -69,7 +69,7 @@ async def try_release(key: CacheKey, configuration_snapshot: CacheConfiguration)
             configuration_snapshot.eviction_strategy().mark_released(key)
             logger.debug('Released cache key %s', key)
             return True
-        except Exception as e:
+        except BaseException as e:
```
Comment (charles-dyfis-net) [inline memoize/wrapper.py]: Hmm. This is a fairly different case -- the only code being invoked here is either internal to memoize or plugins/extensions adding storage backends to it; we shouldn't ever see exceptions thrown by wrapped code down this path.

File: memoize/wrapper.py [hunk 3]
```diff
@@ -81,7 +81,7 @@ async def refresh(actual_entry: Optional[CacheEntry], key: CacheKey,
             if actual_entry is None:
                 logger.debug('As no valid entry exists, waiting for results of concurrent refresh %s', key)
                 entry = await update_status_tracker.await_updated(key)
-                if isinstance(entry, Exception):
+                if isinstance(entry, BaseException):
```
Comment (charles-dyfis-net) [inline memoize/wrapper.py:84]: :+1:

File: memoize/wrapper.py [hunk 4]
```diff
@@ -115,15 +115,15 @@ async def refresh(actual_entry: Optional[CacheEntry], key: CacheKey,
                         asyncio.ensure_future,
                         try_release(to_release, configuration_snapshot)
                     )
-            except Exception as e:
+            except BaseException as e:
```
Comment (charles-dyfis-net) [inline memoize/wrapper.py]: This is another case where we're calling only internal code, so any nonstandard exception would be something coming from the runtime that we aren't expected to handle.

File: memoize/wrapper.py [hunk 5]
```diff
@@ -115,15 +115,15 @@ async def refresh(actual_entry: Optional[CacheEntry], key: CacheKey,
...
                 logger.error("ignoring failure during eviction after successful refresh", exc_info=e)
             finally:
                 return offered_entry
         except asyncio.TimeoutError as e:
             logger.debug('Timeout for %s: %s', key, e)
             update_status_tracker.mark_update_aborted(key, e)
             raise CachedMethodFailedException('Refresh timed out') from e
-        except Exception as e:
+        except BaseException as e:
```
Comment (charles-dyfis-net) [inline memoize/wrapper.py:126]: 👍 -- exceptions from the `await value_future` end up here.

**Review by charles-dyfis-net — APPROVED** [review]

**Pull request merged**